nva/nvafakebios.c

```diff
@@ -80,8 +80,12 @@ int vbios_upload_pramin(int cnum, uint8_t *vbios, int length)
 			cnum, nva_cards[cnum]->chipset.chipset);
 
 	if (nva_cards[cnum]->chipset.card_type >= 0x50) {
-		uint64_t vbios_vram = (uint64_t)(nva_rd32(cnum, 0x619f04) & ~0xff) << 8;
-
+		if (vbios_vram)
+			/* If vbios vram addr was already chosen, save the addr into 0x619f04 
+			 * 0x1 tells it's vram and 0x8 enables it */
+			nva_wr32(cnum, 0x619f04, (vbios_vram >> 8) + 0x09);
```

Comment:
please use | instead of + for bit operations

Reply:
and please mask the value, so that we don't end up writing into the lower bytes: vbios_vram & 0xffffff00

Reply:
for 512 MB RAM if I call:
lookup -ac0 0x619f04 0x1ffe09
PDISPLAY.VGA.ROM_WINDOW => { BASE = 0x1ffe0000 | ENABLE | TARGET = VRAM }

 0x1ffe0000 is the address I want or am I missing something?

Reply:
ohh, it's more of a "safety" thing. the upper 24 bits encode the address, the lower 8 bits are some random options. And we should just make sure to never have random bits set in the lower 8 ones.

It won't happen right now, as the smallest value of ((vbios_vram >> 8)  is 0xe00... maybe by accident when passing a negative size you can end up with weird values though... anyway, this application usually runs as root and this can easily end up messing up system memory if we are not careful enough.

Reply:
This probably would never happen, even -1 gets shifted with zeroes. But I will ad this anyway.